File: OldCs/Problem004.cs
// -----------------------------------------------------------------------
// <copyright file="Problem004.cs" company="">
// TODO: Update copyright text.
// </copyright>
// -----------------------------------------------------------------------

namespace ProjectEulerSolutions
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;

    /// <summary>
    /// TODO: Update summary.
    /// </summary>
    public class Problem004
    {
        public static void Run()
        {
            var maxPalindrome = getPalindromes().Max();
            Console.WriteLine(maxPalindrome);
        }

        private static IEnumerable<int> getPalindromes()
        {
            return Enumerable.Range(1, int.MaxValue).Select(x => x * 11).TakeWhile(x => x < 1000)
                    .SelectMany(m1 => Enumerable.Range(100, 1000 - 99).Select(m2 => m1 * m2))
                    .Where(x => isPalindrome(x));

            //for (int i = 999; i > 99; i--)
            //    for (int j = 999; j > 99; j--)
            //        yield return isPalindrome(i * j) ? i * j : 1;
        }

        private static bool isPalindrome(long theNumber)
        {
            return theNumber.ToString() == Reverse(theNumber.ToString());
        }

        private static string Reverse(string theNumber)
        {
            return theNumber.Reverse().Aggregate(new StringBuilder(), (builder, c) => builder.Append(c)).ToString();
        }
    }
}

File: Problem004.js
exports.run = function () {
    var palindromes = [],
	number=0;

    function isPalindrome(stringer) {
	return stringer === stringer.split("").reverse().join("");
    };

    for(var i=999; i>0 && !number; i--)
	for(var j=999; j>0 && !number; j--)
	    if(isPalindrome((i*j).toString()))
		palindromes.push(i*j);

    for(var i=0; i<palindromes.length; i++)
	if(palindromes[i]>number)
	    number = palindromes[i];

    console.log("number: " + number);
    return number;
};
